test/functorch/test_control_flow.py

```diff
@@ -1251,8 +1309,163 @@ def mul(x: torch.Tensor, y: torch.Tensor):
+        result_exp = _fake_scan(div, init=init, input=x, dim=0, reverse=reverse)
+        self.assertEqual(result, result_exp)
+
+    @requires_cuda
+    @parametrize("reverse", [False, True])
+    @parametrize("compile_mode", ["none", "eager", "compile", "compile_dynamic_shape"])
+    @parametrize("device", [torch.device("cpu"), torch.device("cuda")])
+    # Skipping the scan tests, as scan currently does not support inductor
+    @decorateIf(
+        unittest.skip,
+        lambda params: params["compile_mode"] != "eager"
+        and params["compile_mode"] != "none",
+    )
+    def test_scan_device_dtype(self, reverse, compile_mode, device):
+        def add(x: torch.Tensor, y: torch.Tensor):
+            return x + x, y + y
+
+        torch.compiler.reset()
+        if compile_mode == "compile":
+            scan_fct = torch.compile(scan, fullgraph=True, dynamic=False)
+        elif compile_mode == "compile_dynamic_shape":
+            scan_fct = torch.compile(scan, fullgraph=True, dynamic=True)
+        elif compile_mode == "eager":
+            scan_fct = torch.compile(scan, fullgraph=True, backend="eager")
+        else:
+            scan_fct = scan
+
+        # Check all outputs and carries on the correct device and with torch.float32
+        x = torch.randn(3, 10, 2, device=device)
+        op, init = (add, torch.zeros(1, 10, 2, device=device))
+        result = scan_fct(op, init=init, input=x, dim=0, reverse=reverse)
+        result_exp = _fake_scan(op, init=init, input=x, dim=0, reverse=reverse)
+        self.assertEqual(result, result_exp)
+        self.assertEqual(
+            [[r.device.type for r in res] for res in result],
+            [[device.type for _ in res] for res in result],
+        )
+        self.assertEqual(
+            [[r.dtype for r in res] for res in result],
+            [[torch.float32 for _ in res] for res in result],
+        )
+
+        # Check all outputs and carries on the correct device and
+        # carry.dtype torch.float32 and output.dtype torch.float16
+        x = torch.randn(3, 10, 2, device=device, dtype=torch.float16)
+        op, init = (add, torch.zeros(1, 10, 2, device=device, dtype=torch.float32))
+        result = scan_fct(op, init=init, input=x, dim=0, reverse=reverse)
+        result_exp = _fake_scan(op, init=init, input=x, dim=0, reverse=reverse)
+        self.assertEqual(result, result_exp)
+        self.assertEqual(
+            [[r.dtype for r in res] for res in result],
+            [
+                [torch.float32 for _ in range(len(result[0]))],
+                [torch.float16 for _ in range(len(result[1]))],
+            ],
+        )
+
+        # Check all outputs and carries on the correct device and
+        # carry.dtype torch.int64 and output.dtype torch.float32
+        x = torch.randn(3, 10, 2, device=device, dtype=torch.float32)
+        op, init = (add, torch.zeros(1, 10, 2, device=device, dtype=torch.float16))
```

Comment:
Can we parameterize on dtype like x_dtype, init_dtype? so we don't have list out all the dtypes.

Reply:
I parameterized the test with `dtype` and included most of the dtypes that torch supports. See [here](https://github.com/pytorch/pytorch/blob/cb06c9cf487b748caa807bcc66207ce2f57f55cf/test/functorch/test_control_flow.py#L1494-L1549)